community/planning/service_lifecycle.md

```diff
@@ -0,0 +1,275 @@
+`LifecycleDispatch` trait instead of the `ServiceOrchestrator` directly. When a
+service needs to be added, retired, or purged, each implementation will be
+called. The implementations will check to see if it supports the service type
+that is being updated, acting as a NO-OP if not. It is expected that service
+type will only be handled by one implementation of the `LifecycleDispatch`.
+
+```rust
+pub trait LifecycleDispatch: Send {
+    // prepare and finalize a service
+    fn add_service(
+        &self,
+        circuit_id: &str,
+        service_id: &str,
+        service_type: &str,
+        args: Vec<(String, String)>,
+    ) -> Result<(), InternalError>;
+
+    fn retire_service(
+        &self,
+        circuit_id: &str,
+        service_id: &str,
+        service_type: &str,
+    ) -> Result<(), InternalError>;
+
+    fn purge_service(
+        &self,
+        circuit_id: &str,
+        service_id: &str,
+        service_type: &str,
+    ) -> Result<(), InternalError>;
+
+    fn shutdown_all_services(&self) -> Result<(), InternalError>;
+
+    fn add_stopped_service(
+        &self,
+        circuit_id: &str,
+        service_id: &str,
+        service_type: &str,
+        args: HashMap<String, String>,
+    ) -> Result<(), InternalError>;
+}
+```
+
+The `ServiceOrchestrator` implements the trait directly.
+
+### SyncLifecycleInterface
+
+The `AdminService` expects service creation to be synchronous. To use the
+asynchronous `LifecycleExecutor` the new `SyncLifecycleInterface` will be
+written to implement a `LifecycleDispatch`.
+
+```rust
+impl SyncLifecycleInterface {
+    pub fn new(
+        store: Box<dyn LifecycleStore + Send>,
+        alarm: Box<dyn ExecutorAlarm>,
+        supported_types: Vec<String>,
+        time_to_wait: time::Duration,
+    ) -> Self {
+        SyncLifecycleInterface {
+            store,
+            alarm,
+            supported_types,
+            time_to_wait,
+        }
+    }
+}
+```
+
+For each service operation, `SyncLifecycleInterface` writes the pending service
+to the `LifecycleStore` and then uses an `ExecutorAlarm` to wake up the executor
+for the pending service. The interface will periodically check the lifecycle
+store to see if the service has been updated by the executor, only returning to
+when the service has been updated or a timeout has been reached.
+
+![Lifecycle Integration Admin Service]({% link images/admin_lifecycle.svg %}
+  "Lifecycle Integration with Admin Service"){:.centered}
+
+Due to the difference in designs, two of the required methods on the
+`LifecycleDispatch` are NO-OPs for the `SyncLifecycleInterface`. Services do not
+need to be stopped and restarted on shutdown. Therefore `shutdown_all_services`
+and `add_stopped_service` simply return `Ok(())`, same for if `add_service` is
+called and if the service was already added.
+
+`SyncLifecycleInterface` combines both `prepare_service` and `finish_service` to
```

Comment:
is `finish_service` supposed to be `finalize_service`?

Reply:
fixed